no-loop-func warnings at ForInStatement.right and ForStatement.init.

I got `no-loop-func` warnings at this code:

```js
"use strict";
let xs = [1, 2, 3];
for (let x of xs.filter(x0 => x0 != null)) {
  //                    ^^ here!
  console.log(x);
}
for (let x in xs.filter(x0 => x0 != null)) {
  //                    ^^ here!
  console.log(x);
}
for (let i = 0, ys = xs.filter(x0 => x0 != null); i < ys.length; i++) {
  //                           ^^ here!
  console.log(ys[i]);
}
```

But I think that `ForInStatement.right` and `ForStatement.init` are out of loop, since those run once at first.

I'm happy to make PR if this is accepted :)

----

version: `v0.20.0`
.eslintrc:

```
{
  "env": {
    "node": true,
    "es6": true
  },
  "ecmaFeatures": {
    "arrowFunctions": true,
    "forOf": true
  }
}
```

Sorry, I had checked on Sublime Text.

```
test.js
   3:24  error  Don't make functions within a loop  no-loop-func
   7:24  error  Don't make functions within a loop  no-loop-func
  11:31  error  Don't make functions within a loop  no-loop-func

✖ 3 problems (3 errors, 0 warnings)
```

I expect there aren't errors.

--- a/lib/rules/no-loop-func.js
+++ b/lib/rules/no-loop-func.js
@@ -11,23 +11,6 @@
 //------------------------------------------------------------------------------
 
 module.exports = function(context) {
-    var loopNodeTypes = [
-        "ForStatement",
-        "WhileStatement",
-        "ForInStatement",
-        "ForOfStatement",
-        "DoWhileStatement"
-    ];
-
-    /**
-     * Checks if the given node is a loop.
-     * @param {ASTNode} node The AST node to check.
-     * @returns {boolean} Whether or not the node is a loop.
-     */
-    function isLoop(node) {
-        return loopNodeTypes.indexOf(node.type) > -1;
-    }
-
     /**
      * Reports if the given node has an ancestor node which is a loop.
      * @param {ASTNode} node The AST node to check.
@@ -36,7 +19,31 @@ module.exports = function(context) {
     function checkForLoops(node) {
         var ancestors = context.getAncestors();
 
-        if (ancestors.some(isLoop)) {
+        /**
+         * Checks if the given node is a loop and current context is in the loop.
+         * @param {ASTNode} ancestor The AST node to check.
+         * @param {number} index The index of ancestor in ancestors.
+         * @returns {boolean} Whether or not the node is a loop and current context is in the loop.
+         */
+        function isInLoop(ancestor, index) {
+            switch (ancestor.type) {
+                case "ForStatement":
+                    return ancestor.init !== ancestors[index + 1];
+
+                case "ForInStatement":
+                case "ForOfStatement":
+                    return ancestor.right !== ancestors[index + 1];
+
+                case "WhileStatement":
+                case "DoWhileStatement":
+                    return true;
+
+                default:
+                    return false;
+            }
+        }
+
+        if (ancestors.some(isInLoop)) {
             context.report(node, "Don't make functions within a loop");
         }
     }

--- a/tests/lib/rules/no-loop-func.js
+++ b/tests/lib/rules/no-loop-func.js
@@ -23,7 +23,13 @@ var eslintTester = new ESLintTester(eslint),
 eslintTester.addRuleTest("lib/rules/no-loop-func", {
     valid: [
         "string = 'function a() {}';",
-        "for (var i=0; i<l; i++) { } var a = function() { };"
+        "for (var i=0; i<l; i++) { } var a = function() { };",
+        "for (var i=0, a=function() { }; i<l; i++) { }",
+        "for (var x in xs.filter(function(x) { return x != null; })) { }",
+        {
+            code: "for (var x of xs.filter(function(x) { return x != null; })) { }",
+            ecmaFeatures: { forOf: true }
+        }
     ],
     invalid: [
         {
@@ -51,6 +57,14 @@ eslintTester.addRuleTest("lib/rules/no-loop-func", {
         {
             code: "for (var i=0; i < l; i++) { function a() {}; a(); }",
             errors: [ { message: expectedErrorMessage, type: "FunctionDeclaration" } ]
+        },
+        {
+            code: "for (var i=0; (function() {})(), i<l; i++) { }",
+            errors: [ { message: expectedErrorMessage, type: "FunctionExpression" } ]
+        },
+        {
+            code: "for (var i=0; i<l; (function() {})(), i++) { }",
+            errors: [ { message: expectedErrorMessage, type: "FunctionExpression" } ]
         },
         {
             code: "while(i) { (function() {}) }",

